secrets/store.go

```diff
@@ -27,21 +38,43 @@ type Store struct {
 //
 // Context should come with a timeout otherwise this might block forever, i.e.
 // if the path never becomes available.
-func NewStore(ctx context.Context, path string, logger log.Wrapper) (*Store, error) {
-	parser := func(r io.Reader) (interface{}, error) {
-		secrets, err := NewSecrets(r)
-		if err != nil {
-			return nil, err
-		}
-		return secrets, nil
+func NewStore(ctx context.Context, path string, logger log.Wrapper, middlewares ...SecretMiddleware) (*Store, error) {
+	store := &Store{}
+	if len(middlewares) > 0 {
+		store.SecretHandler(middlewares...)
 	}
-	result, err := filewatcher.New(ctx, path, parser, logger)
+
+	result, err := filewatcher.New(ctx, path, store.parser, logger)
 	if err != nil {
 		return nil, err
 	}
-	return &Store{
-		watcher: result,
-	}, nil
+
+	store.watcher = result
+	return store, nil
+}
+
+func (s *Store) parser(r io.Reader) (interface{}, error) {
+	secrets, err := NewSecrets(r)
+	if err != nil {
+		return nil, err
+	}
+
+	if s.secretHandlerFunc != nil {
+		s.secretHandlerFunc(secrets)
+	}
+
+	return secrets, nil
+}
+
+// SecretHandler creates the middleware chain.
+func (s *Store) SecretHandler(middlewares ...SecretMiddleware) {
```

Comment:
Is there a reason to export this function? I thought just allow the callers to specify middlewares in `NewStore` should be enough?

Reply:
My mind isn't fully made up. I think in general it's a good idea to allow additional middleware gets attached for some unforeseen reason, but I'm happy to make it private until we have an actual use case.